llvm/lib/Transforms/Vectorize/LoopVectorize.cpp

```diff
@@ -9014,9 +9014,11 @@ LoopVectorizationPlanner::tryToBuildVPlanWithVPRecipes(VFRange &Range) {
 
   DebugLoc DL = getDebugLocFromInstOrOperands(Legal->getPrimaryInduction());
   TailFoldingStyle Style = CM.getTailFoldingStyle(IVUpdateMayOverflow);
-  // When not folding the tail, we know that the induction increment will not
-  // overflow.
-  bool HasNUW = Style == TailFoldingStyle::None;
+  // Use NUW for the induction increment if we proved that it won't overflow in
+  // the vector loop or when not folding the tail. Then we know that the
+  // induction increment will not overflow as the vector trip count is >=
+  // increment and a multiple of the increment.
+  bool HasNUW = !IVUpdateMayOverflow || Style == TailFoldingStyle::None;
```

Comment:
I think (although I'm not sure) you can drop the `Style == TailFoldingStyle::None`, as I think `!IVUpdateMayOverflow` subsumes it.

Reply:
It is still needed, as `IVUpdateMayOverflow` can only handle a subset of statically known cases; when not folding the tail, the runtime checks guarantee that there is no overflow.